### PR TITLE
[threads] Add back mono_threads_attach_tools_thread as a public API

### PR DESCRIPTION
**Pull request by lambdageek**
In https://github.com/mono/mono/commit/a5da7b21f4b6dbc5eaa09c2addee91b84dc1dbd5
we got rid of "tools" threads internally to the runtime.

However since the API was previously marked with MONO_API it was an internal
API that some embedders depended on.

This PR adds back an (external-only) limited form of tools thread.

The runtime is aware of the Tools thread in that FOREACH_THREAD_* macros will
iterate over them, and the thread has a coop thread state machine. (That is,
mono_thread_info_current() and GC Safe and GC Unsafe transitions all work.)

However the thread is:
1. Not stopped by the GC
2. Is not interrupted by profiler sampling.
3. Does not have a "current domain"
4. (As a consequence of the above) cannot call managed methods or touch managed
objects.

Such threads are useful for low-level interaction with the runtime such as
querying metadata, the JIT state and other coordination.

mono_threads_attach_tools_thread should be called no more than once.  It should
not be called on a thread that is already attached with mono_thread_atach, and
vice versa.

Addresses https://github.com/mono/mono/issues/18011

### CODE REVIEW

**Comment by lambdageek** [comment]
This is enough to get the sample from https://github.com/lambdageek/mono-mach-exceptions working.

Another option is to do stuff with mach ports in Mono directly and provide a mach exception hook that an embedder can register.  We'd probably have to expose a function like:
```
  typedef ... (*mono_mach_exception_hook_fn)(...);
  void mono_mach_add_exception_hook (mono_mach_exception_hook_fn fnptr);
```

where `mono_mach_exception_hook_fn` would be a signature like that of [`catch_exception_raise` from Mach](http://web.mit.edu/darwin/src/modules/xnu/osfmk/man/catch_exception_raise.html)

**Comment by alexrp** [comment]
Is it really a good idea to take a stance of supporting legacy APIs just because they happened to be exported from `libmono.so`? It's not like `mono-threads.h` is a public header. Many other symbols (`mono_...`, `ves_icall_...`, etc) are incidentally exported from `libmono.so` but that doesn't mean people are supposed to be accessing/using them.

It seems to me that if these APIs are needed by embedders, then they should be properly promoted to an actual public header with the support commitment that goes with that.

**Comment by lambdageek** [comment]
> It seems to me that if these APIs are needed by embedders, then they should be properly promoted to an actual public header with the support commitment that goes with that.

That's true.

**Comment by lambdageek** [comment]
Made is a public API

File: mono/metadata/threads.c [hunk 1]
```diff
@@ -1550,6 +1550,35 @@ mono_thread_attach (MonoDomain *domain)
...
+ * a "current domain" because it cannot run managed code or interact with
+ * managed objects.  However it can act on some metadata, and use our low-level
+ * locks and the coop thread state machine (ie GC Safe and GC Unsafe
+ * transitions make sense).
+ */
+void
+mono_threads_attach_tools_thread (void)
+{
+	MonoThreadInfo *info;
+	info = mono_thread_info_attach ();
```
Comment (jaykrell) [inline mono/metadata/threads.c]: Combine with previous line?

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by lambdageek** [comment]
@monojenkins backport to 2019-12

**Comment by lambdageek** [comment]
@monojenkins backport to 2019-10

**Pull request merged**